ngraph/core/reference/src/runtime/reference/tile.cpp

```diff
@@ -14,34 +14,100 @@
 // limitations under the License.
 //*****************************************************************************
 
+#include <algorithm>
 #include <cmath>
-#include <stdio.h>
+#include <cstdio>
+#include <numeric>
 
 #include "ngraph/check.hpp"
 #include "ngraph/runtime/reference/tile.hpp"
 
 using namespace ngraph;
 
-void runtime::reference::tile(
-    const char* arg, char* out, const Shape& in_shape, const Shape& out_shape, size_t elem_size)
+namespace
+{
+    /// \brief For each axis calculates the product of inner axes
+    /// If dims has shape (2, 3, 4) then for 2 (first axis) the inner axes would be (3, 4)
+    /// and for 3 (second axis) it would be (4)
+    /// If dims has shape(2, 3, 4) then the output vector would be (3 * 4, 4, 1)
+    /// The outermost axis is not used. For innermost axis it is always 1.
+    /// \param[in] dims Shape of the output
+    ///
+    /// \return Vector containing calculated values for each axis.
+    std::vector<int64_t> create_pitches(const Shape& dims)
+    {
+        std::vector<int64_t> pitch;
+        pitch.resize(dims.size());
+        std::partial_sum(dims.rbegin(), dims.rend() - 1, pitch.begin(), std::multiplies<int64_t>());
+        std::reverse(pitch.begin(), pitch.end() - 1);
+
+        pitch.back() = 1;
+        return pitch;
+    }
+
+    bool calculate_next_axis(std::vector<int64_t>& indices, int64_t& axis, const Shape& shape)
+    {
+        if (axis-- == 0)
+        {
+            return false;
+        }
+
+        if (++indices[axis] != shape[axis])
+        {
+            axis = indices.size();
+            return false;
+        }
+
+        indices[axis] = 0;
+        return true;
+    }
+}
+
+void runtime::reference::tile(const char* arg,
+                              char* out,
+                              const Shape& in_shape,
+                              const Shape& out_shape,
+                              const size_t elem_size,
+                              const std::vector<int64_t>& repeats)
 {
     Shape in_shape_expanded(in_shape);
     in_shape_expanded.insert(in_shape_expanded.begin(), out_shape.size() - in_shape.size(), 1);
-    CoordinateTransform input_transform(in_shape_expanded);
-    CoordinateTransform output_transform(out_shape);
+    size_t block_size = 0;
+    int64_t num_repeats = 0;
+    const int input_rank = in_shape_expanded.size();
+    const int64_t last_dim = in_shape_expanded[input_rank - 1];
+    const std::vector<int64_t> pitches = create_pitches(out_shape);
+    const char* copy = nullptr;
 
-    for (const Coordinate& output_coord : output_transform)
+    std::vector<int64_t> indices(in_shape_expanded.size() - 1, 0);
+    int64_t axis = indices.size();
+
+    while (axis >= 0)
```

Comment:
If you use `size_t` for axis (because indices.size() returns `size_t`) I think you can use the conditional: `axis <= indices.size();`